Coverage for in-app Payments

AndreiH opened the pull request.

AndreiH commented:
Commits need to be rebased before :)

bobsilverberg commented on marketplacetests/tests/test_marketplace_make_an_in-app_payment.py:
```diff
@@ -0,0 +1,71 @@
+        self.assertTrue('Confirm payment', payment.confirm_payment_panel_text)
+        self.assertTrue('test 0.10', payment.in_app_product_name)
+
+        payment.tap_buy_button()
```
The test fails for me at this point:

```
TEST-START test_marketplace_make_an_in-app_payment.py
test_make_an_in_app_payment (test_marketplace_make_an_in-app_payment.TestMakeInAppPayment) ... ERROR

======================================================================
ERROR: None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette_test.py", line 170, in run
    testMethod()
  File "/Users/bsilverberg_local/gitRepos/marketplace-tests-gaia/marketplacetests/tests/test_marketplace_make_an_in-app_payment.py", line 69, in test_make_an_in_app_payment
    payment.tap_buy_button()
  File "/Users/bsilverberg_local/gitRepos/marketplace-tests-gaia/marketplacetests/payment/app.py", line 94, in tap_buy_button
    self.marionette.find_element(*self._buy_button_locator).tap()
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 1213, in find_element
    response = self._send_message('findElement', 'value', **kwargs)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/decorators.py", line 35, in _
    return func(*args, **kwargs)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 633, in _send_message
    self._handle_error(response)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 664, in _handle_error
    raise NoSuchElementException(message=message, status=status, stacktrace=stacktrace)
TEST-UNEXPECTED-FAIL | test_marketplace_make_an_in-app_payment.py test_marketplace_make_an_in-app_payment.TestMakeInAppPayment.test_make_an_in_app_payment | NoSuchElementException: NoSuchElementException: Unable to locate element: //button[text()='Buy']

----------------------------------------------------------------------
Ran 1 test in 157.829s

FAILED (errors=1)
```

AndreiH replied:
Ups, this happens because I am using <code>buy</code> button from payments.
I forgot that there are two separate frames for in-app payments and simple payments.
I will get this fixed! Thanks!

bobsilverberg commented:
Thanks for this @AndreiH, great work! I'm going to go through and do a thorough review of this, but one general comment I wanted to add is that we shouldn't put dashes `-` into file names. Please change the dashes in the file names to underscores.

bobsilverberg commented on marketplacetests/firefox_accounts/app.py:
```diff
@@ -18,15 +21,19 @@ class FirefoxAccounts(Base):
-        self.marionette.switch_to_frame(self.frame)
-        self.wait_for_element_not_present(*self._body_loading_locator)
-        self.wait_for_element_displayed(*self._firefox_logo_locator)
+        time.sleep(1)
```
We really try to avoid using `sleep`. Can you define a wait that will work instead, e.g., wait for either frame to be present?

AndreiH replied:
done

bobsilverberg commented on marketplacetests/firefox_accounts/app.py:
```diff
@@ -37,6 +44,20 @@ def login(self, email, password):
+        self.marionette.switch_to_frame()
+        self.wait_for_element_present(*self._ff_accounts_frame_locator)
+        self.frame = self.marionette.find_element(*self._ff_accounts_frame_locator)
+        self.marionette.switch_to_frame(self.frame)
```
These four lines of code (54-57) are basically the same as the ones in `switch_to_in_app_login_frame` (lines 48-51), with the only difference being the locator used. I suggest making them a single method that accepts a locator as an argument, and then that method can be called when needed.

AndreiH replied:
done

bobsilverberg commented on marketplacetests/firefox_accounts/app.py:
```diff
@@ -50,5 +71,20 @@ def tap_next(self):
 
     def tap_sign_in(self):
         self.keyboard.dismiss()
+        self.change_to_correct_login_frame()
```
This is the second time this is being called (it's also called in `__init__` at https://github.com/mozilla/marketplace-tests-gaia/pull/70/files#diff-4748f41d7933badeadce5c5003aaf2acR28), and it includes a test to check which frame to switch to. I suggest we store the information about which frame to switch to the first time this is called, and then the next time all we'll have to do is switch to that frame.

AndreiH replied:
I agree with this, it would be an improvement.
I tried saving the current frame after we switch it and returning that. But it switches me to the app that triggered the log in and not the log in frame :|
Do you have any suggestions on how to implement this? Thanks!

bobsilverberg replied:
I wouldn't save the frame after switching to it, rather I would keep the same logic that is used now to figure out which frame to switch to (which is done before switching), and save that information.

bobsilverberg commented on marketplacetests/firefox_accounts/app.py:
```diff
@@ -50,5 +71,20 @@ def tap_next(self):
+    def tap_not_you(self):
+        self.marionette.find_element(*self._not_you_logout_link_locator).tap()
+
+    def wait_password_field_visible(self):
```
Suggest changing to `wait_for_password_field_visible`.

AndreiH replied:
done

bobsilverberg commented on marketplacetests/marketplace/regions/app_details.py:
```diff
@@ -28,6 +28,10 @@ def install_button_text(self):
         return self.marionette.find_element(*self._install_button_locator).text
 
     @property
+    def install_button_class(self):
```
This is not used anywhere.

AndreiH replied:
removed

bobsilverberg commented on marketplacetests/marketplace/regions/search_results.py:
```diff
@@ -25,6 +23,12 @@ def search_results(self):
         return [Result(self.marionette, result) for result in search_results]
 
+    def install_searched_app(self, search_term):
+        # Select the application link in the list
```
I don't think these comments are necessary. The second one doesn't make much sense to me in the context of the code. The method itself is self-explanatory.

AndreiH replied:
removed the comments

bobsilverberg commented on marketplacetests/marketplace/regions/search_results.py:
```diff
@@ -25,6 +23,12 @@ def search_results(self):
         search_results = self.marionette.find_elements(*self._search_result_locator)
         return [Result(self.marionette, result) for result in search_results]
 
+    def install_searched_app(self, search_term):
```
Also, I'm not sure we need this at all. We not use the existing `navigate_to_app` which is already used by lots of other tests to navigate to the app in question, and then install from there?

AndreiH replied:
I agree with that, but I also think it is better for us to have coverage with both installing methods.
Users will search for the apps and install them directly from the search page.
Do you agree with having both?

bobsilverberg replied:
If that is a case we need covered, it should have its own test. The purpose of this test if to test in-app payments, not whether we can install directly from search results, so I think we should remove it from here and check with @krupa to see if we need another test for installing directly from search results.

AndreiH replied:
I will make the required changed then. Remove this extra method

bobsilverberg commented on marketplacetests/marketplace_gaia_test.py:
```diff
@@ -29,5 +29,5 @@ def install_certs(self):
                                   destination='data/b2g/mozilla/%s/%s' % (profile_folder, file_name))
         self.data_layer.set_char_pref('dom.mozApps.signed_apps_installable_from',
-                                      'https://marketplace-dev.allizom.org,https://marketplace.firefox.com')
+                                      'https://marketplace-dev.allizom.org,https://marketplace.firefox.com,https://marketplace.allizom.org')
```
Is there a reason we need this? Aren't we only installing from -dev?

AndreiH replied:
I added this because in-app tester App was only added on stage and then added on dev and I forgot about it.
I will remove it.

bobsilverberg commented on marketplacetests/payment/app.py:
```diff
@@ -39,6 +42,14 @@ def app_name(self):
+        return self.marionette.find_element(*self._in_app_product_name_locator).text
+
+    @property
+    def confirm_payment_panel_text(self):
```
I'm not sure why the property is called `confirm_payment_panel_text` but the locator is called `_confirm_payment_header_locator`. This is confusing. I suggest being consistent with the names, whichever one is more appropriate.

AndreiH replied:
:+1:

bobsilverberg commented on marketplacetests/payment/regions/in_app.py:
```diff
@@ -0,0 +1,35 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
```
From what I can tell, this is an app object for the "Testing In-App-Payments" test app. If that is correct then I don't think it really belongs here under `marketplacetests/payment/regions/` as it is not a part of the payment system.  I suggest putting it elsewhere.

bobsilverberg replied:
Also, name it differently and add a comment describing what it is. If I am correct in my assumption above, that is.

AndreiH replied:
True, I will find another place for it and describe it better.

bobsilverberg commented on marketplacetests/tests/test_marketplace_in-app_not_you_login_feature.py:
```diff
@@ -0,0 +1,71 @@
+    APP_NAME = 'Testing In-App-Payments'
+    APP_TITLE = 'In-App-Payments'
+
+    def setUp(self):
```
This `setUp` is quite lengthy and is duplicated in here and in `test_marketplace_make_an_in-app_payment.py` I suggest moving it into `MarketplaceGaiaTestCase` and giving it a specific name, like `install_in_app_payments_test_app`, and then calling that from each `setUp`.

AndreiH replied:
:+1:  good idea!

bobsilverberg commented on marketplacetests/tests/test_marketplace_in-app_not_you_login_feature.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_make_an_in_app_payment(self):
+
+        username = self.testvars['marketplace']['username']
```
I suggest changing the test to use a new user.

AndreiH replied:
I had problems with the payments with a new user. Payments not accepted in that region.
This is because the newly created user has to have USA region selected for the payments to complete. That means an  extra step to the test and also creating a pin. I don't have a problem with adding it, seems fair to have a new user every time. 
And we don't need a PIN added to the credentials file. I will update this accordingly.

bobsilverberg commented on marketplacetests/tests/test_marketplace_in-app_not_you_login_feature.py:
```diff
@@ -0,0 +1,71 @@
+        payment = Payment(self.marionette)
+        payment.tap_cancel_pin()
+
+        tester_app.tap_buy_product()
```
This is confusing. In `tap_buy_product`, there is an argument called `is_logged_in` which defaults to `True`, and if so then returns `Payment`, otherwise it returns `FirefoxAccounts`. This is the only instance I can see where `tap_buy_product` is called and `is_logged_in` would be `True`, yet the code does not seem to expect `Payment` to be returned. In fact it does not interact with `Payment` at all from this point on, it only interacts with `fxa`.

This suggests there is something invalid/unnecessary with the `tap_buy_product` method.

AndreiH replied:
true, we are not returning Payments in any test for now. Removed it!

bobsilverberg commented on marketplacetests/tests/test_marketplace_make_an_in-app_payment.py:
```diff
@@ -0,0 +1,71 @@
+    APP_NAME = 'Testing In-App-Payments'
+    APP_TITLE = 'In-App-Payments'
+
+    def setUp(self):
```
See my note above regarding this `setUp`.

bobsilverberg commented on marketplacetests/tests/test_marketplace_make_an_in-app_payment.py:
```diff
@@ -0,0 +1,71 @@
+
+    def test_make_an_in_app_payment(self):
+
+        email = self.testvars['marketplace']['username']
```
As above, suggest using a new user.

AndreiH replied:
So, yes payment cannot be done using a new account because region needs to be changed to USA before completing the payment.
This requires more extra steps (creating pin, opening Marketplace, log in there and start change region process) and then switch back to payment frame. @bobsilverberg do you thinks that it's worth adding all these extra steps?
![2015-01-13-19-05-06](https://cloud.githubusercontent.com/assets/2787766/5724917/84ec713a-9b57-11e4-8aac-b576e22a8c3b.png)

AndreiH replied:
There is another issue with this. @krupa requested to test in-app Payments on Stage, and we are testing Marketplace Dev. So in order to make those extra steps happen, we have to modify in-app tests to open Marketplace Stage, login with new user, change region.
I will ask @krupa about this, if in-app payments tests should run on stage API

bobsilverberg commented:
Great work @AndreiH. I've added quite a lot of comments, but in general this is an excellent PR. Thanks for the new coverage. Let me know once my comments have been addressed and I'll review it again.

AndreiH commented:
Thanks for reviewing this @bobsilverberg . I will upload an update soon!

AndreiH commented on marketplacetests/marketplace/regions/app_details.py:
```diff
@@ -45,13 +45,6 @@ def tap_write_review(self, logged_in=True):
             return AddReview(self.marionette)
 
-    def tap_purchase_button(self, is_logged_in=True):
+    def tap_install_button(self):
```
I modified this because we are now using it to install free apps as well.

AndreiH commented:
@bobsilverberg I think I addressed all your comments. Except the one with creating a new user for every run.

AndreiH commented:
Got my answer from @krupa , we should run in-app Payments tests on Dev so I have to update the test app on dev, add products. If you agree with the extra steps listed in my replies @bobsilverberg , I will begin updating. Thanks!

AndreiH commented:
@bobsilverberg finished updating this! :)

bobsilverberg commented on marketplacetests/marketplace_gaia_test.py:
```diff
@@ -18,6 +23,40 @@ def setUp(self):
+
+            # Confirm the installation and wait for the app icon to be present
+            confirm_install = ConfirmInstall(self.marionette)
+            confirm_install.tap_confirm()
```
This process is failing for me here.

```
TEST-START test_marketplace_make_an_in_app_payment.py
test_make_an_in_app_payment (test_marketplace_make_an_in_app_payment.TestMakeInAppPayment) ... ERROR

======================================================================
ERROR: None
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette_test.py", line 152, in run
    self.setUp()
  File "/Users/bsilverberg_local/gitRepos/marketplace-tests-gaia/marketplacetests/tests/test_marketplace_make_an_in_app_payment.py", line 22, in setUp
    self.install_in_app_payments_test_app()
  File "/Users/bsilverberg_local/gitRepos/marketplace-tests-gaia/marketplacetests/marketplace_gaia_test.py", line 39, in install_in_app_payments_test_app
    confirm_install.tap_confirm()
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/gaiatest/apps/homescreen/regions/confirm_install.py", line 14, in tap_confirm
    self.wait_for_element_displayed(*self._confirm_install_button_locator)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/gaiatest/apps/base.py", line 44, in wait_for_element_displayed
    lambda m: m.find_element(by, locator).is_displayed())
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/wait.py", line 143, in until
    cause=last_exc)
TimeoutException: Traceback (most recent call last):
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/wait.py", line 122, in until
    rv = condition(self.marionette)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/gaiatest/apps/base.py", line 44, in <lambda>
    lambda m: m.find_element(by, locator).is_displayed())
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 1213, in find_element
    response = self._send_message('findElement', 'value', **kwargs)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/decorators.py", line 35, in _
    return func(*args, **kwargs)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 633, in _send_message
    self._handle_error(response)
  File "/Users/bsilverberg_local/.virtualenvs/marketplace-tests-gaia/lib/python2.7/site-packages/marionette/marionette.py", line 664, in _handle_error
    raise NoSuchElementException(message=message, status=status, stacktrace=stacktrace)
TEST-UNEXPECTED-FAIL | test_marketplace_make_an_in_app_payment.py test_marketplace_make_an_in_app_payment.TestMakeInAppPayment.test_make_an_in_app_payment | TimeoutException: Traceback (most recent call last):

----------------------------------------------------------------------
Ran 1 test in 151.126s

FAILED (errors=1)
```

AndreiH replied:
Damn, that's because I also made a change in <code>from gaiatest.apps.homescreen.regions.confirm_install import ConfirmInstall</code>, but it didn't get through

viorelaioia commented on marketplacetests/firefox_accounts/app.py:
```diff
@@ -50,5 +60,20 @@ def tap_next(self):
+
+    @property
+    def signed_in_as_email(self):
+        return self.marionette.find_element(*self._email_input_locator).text
```
can we rename this method to <code>email</code> only?

AndreiH commented:
This has been updated. @bobsilverberg r? Thanks!

AndreiH commented:
I updated the tests to master, @bobsilverberg can you r? Thanks!

bobsilverberg commented:
@AndreiH This looks good, and one of the tests is passing. As discussed in IRC, if you can xfail `test_marketplace_make_an_in_app_payment` with the bug you opened this should be good to merge. Please feel free to squash the commits along with that xfail.

AndreiH commented:
I just updated the tests, squashed the commits. Bug https://bugzilla.mozilla.org/show_bug.cgi?id=1132488 was fixed so no need to xfail.
And I also unxfailed the Payment test and started an adhoc run here:
http://jenkins1.qa.scl3.mozilla.com/view/Marketplace/job/flame-kk-319.mozilla-b2g32_v2_0.nightly.ui.marketplace-tests.adhoc/43/console

bobsilverberg commented:
Looks good. Thanks @AndreiH!
